GVRf/Framework/framework/src/main/java/org/gearvrf/x3d/AnimationInteractivityManager.java

```diff
@@ -628,17 +663,122 @@ public void onSensorEvent(SensorEvent event) {
+                                            GVRSceneObject inlineSceneObject = inlineObject.getInlineGVRSceneObject();
+                                            GVRAnimator animator = (GVRAnimator) inlineSceneObject.getComponent(GVRAnimator.getComponentType());
+                                            if ( animator != null ) {
+                                                List<GVRAnimation> gvrAnimations = animator.getAnimationList();
+                                                GVRAnimation gvrAnimation = null;
+
+                                                if  ( (inlineObject.getBeginKeyFrame() == InlineObject.UNSET_KEY_FRAME) || (inlineObject.getEndKeyFrame() == InlineObject.UNSET_KEY_FRAME)) {
+                                                    if ( (inlineObject.getBeginKeyFrame() == InlineObject.UNSET_KEY_FRAME) && (inlineObject.getEndKeyFrame() == InlineObject.UNSET_KEY_FRAME)) {
+                                                        // use the original FBX animation.
+                                                        gvrAnimation = gvrAnimations.get(0);
+                                                    }
+                                                    else {
+                                                        Log.e(TAG, "Error: ROUTE specifying the keyFrame range - only 1 may be an integers");
+                                                    }
+                                                }
+                                                else {
+                                                    // FBX animation specified in ROUTE as a set of keyFrames
+                                                    // play only a subset of the animation based on keyframes
+                                                    gvrAnimation = interactiveObjectFinal.getFBXAnimation();
+                                                    // check if we already copied over the animation
+                                                    if (gvrAnimation == null) {
+                                                        GVRAnimation gvrOriginalAnimation = gvrAnimations.get(0);
+                                                        if (gvrOriginalAnimation instanceof GVRKeyFrameAnimation) {
+                                                            GVRKeyFrameAnimation gvrOriginalKeyFrameAnimation = (GVRKeyFrameAnimation) gvrOriginalAnimation;
+                                                            GVRAnimationChannel gvrOriginalAnimationChannel = gvrOriginalKeyFrameAnimation.getChannel(0);
+
+                                                            try {
+                                                                // copy the original animation, then save just the requested frames
+                                                                // and their translation, scale and rotate
```

Comment:
We don't have a function to call GVRAnimation or GVRAnimator to begin and end at specific key frames.  Currently, we can only begin at key frame 0 and ends at the end of the animation.
However, selecting from one key frame to another would be a great feature for GearVR.  3d artists often build character animations on a single mesh, and to go from specific key frames on the same mesh.  For example, 'running' would be key frame 0 to 1, 'crawling' could be keyframe 1 to 2, etc.  Currently, GearVR doesn't allow selecting from one specific key frame to another.
The feature I wrote was specific to control FBX animations from X3D, a non-standard, non-documented feature requested by our art team.
We should discuss this to see if we want to instead modify GVRAnimation to see if we can implement the ability to select an animation from specific key frames.